File: Array_And_Strings/5_oneAway.cpp
// there are 3 operations : insert a character, delete a character, replace a character. 
// Given 2 strings, find whether one needs 1 or 0 operations to convert to the other.

#include<iostream>
using namespace std;

bool isOneAway(char* a, char* b, int m, int n)
{
	if(m-n>=2)
		return false;
	int p1=0, p2=0;
	while(p1<m && p2<n)
	{
		if(a[p1]==b[p2])
		{
			p1++;
			p2++;
		}
		else if(p2!=p1)
			return false;
		else if(m>n)
			p1++;
		else 
			p2++;
	}

	return true;
}

int main()
{
	int m,n;
	cout<<"Enter the length of the first string : ";
	cin>>m;
	char a[m];
	cout<<"Enter the string : ";
	cin>>a;
	a[m] = '\0';
	cout<<"Enter the length of the second string : ";
	cin>>n;
	char b[n];
	cout<<"Enter the string : ";
	cin>>b;
	b[n] = '\0';

	if(isOneAway(a,b,m,n))
		cout<<"One away\n";
	else
		cout<<"More than one away\n";

	return 0;
}

File: Array_And_Strings/5_oneAway.java
// Given 2 strings, check if they can be made qual with at most 1 operation
// Operations : Insert , delete, replace

import java.util.Scanner;

class OneAway {

	public boolean check(String a, String b) {
		int l1 = a.length();
		int l2 = b.length();

		if(Math.abs(l1-l2) >= 2)
			return false;

		int i1 = 0;
		int i2 = 0;
		int changes = 0;

		while(i1<l1 && i2<l2) {
			if(changes>=2) {
				return false;
			}
			else if (a.charAt(i1)==b.charAt(i2)) {
				i1++;
				i2++;
			} 
			else if (l1==l2) {
				i1++;
				i2++;
				changes++;
			} 
			else if (l1>l2) {
				i1++;
				changes++;
			}
			else {
				i2++;
				changes++;
			}
		}

		return changes<2;
	}

	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		System.out.print("Enter the 1st string : ");
		String a = in.nextLine();
		System.out.print("Enter the 2nd string : ");
		String b = in.nextLine();

		OneAway ob = new OneAway();
		if(ob.check(a, b))
			System.out.println("One away");
		else 
			System.out.println("More than one away");
	}
}
